storage/grpc_client.go

```diff
@@ -1692,31 +1709,63 @@ func (w *gRPCWriter) uploadBuffer(recvd int, start int64, doneReading bool) (*st
 		}
 
 		// The buffer has been uploaded and there is still more data to be
-		// uploaded, but this is not a resumable upload session. Therefore
-		// keep the stream open and don't commit yet.
-		if !finishWrite && w.chunkSize == 0 {
-			return nil, offset, false, nil
+		// uploaded, but this is not a resumable upload session. Therefore,
+		// don't check persisted data.
+		if !lastWriteOfEntireObject && w.chunkSize == 0 {
+			return nil, writeOffset, nil
 		}
 
-		// Done sending data. Close the stream to "commit" the data sent.
-		resp, finalized, err := w.commit()
+		// Done sending data (remainingDataFitsInSingleReq should == true if we
+		// reach this code). Receive from the stream to confirm the persisted data.
+		resp, err := w.stream.Recv()
 		// Retriable errors mean we should start over and attempt to
 		// resend the entire buffer via a new stream.
 		// If not retriable, falling through will return the error received
 		// from closing the stream.
 		if shouldRetry(err) {
-			sent = 0
-			finishWrite = false
-			offset, err = w.determineOffset(start)
-			if err == nil {
-				continue
+			writeOffset, err = w.determineOffset(start)
+			if err != nil {
+				return nil, 0, err
 			}
+			sent = int(writeOffset) - int(start)
+			continue
 		}
 		if err != nil {
-			return nil, 0, false, err
+			return nil, 0, err
 		}
 
-		return resp.GetResource(), offset, finalized, nil
+		// Confirm the persisted data
+		if !lastWriteOfEntireObject {
+			if resp.GetPersistedSize() != writeOffset {
+				// retry
+				writeOffset = resp.GetPersistedSize()
+				sent = int(writeOffset) - int(start)
+				continue
+			}
+		} else {
+			// If the object is done uploading, close the send stream and check for errors.
+			err = w.stream.CloseSend()
+			if err != nil {
```

Comment:
Is this not also potentially a retry case depending on the error?

Reply:
I see, that is kind of confusing and seems like a strange contract for the method to return an error but never send a non-nil one. Maybe just make a note in a comment about this then?

Reply:
It looks like they are leaving the possibility open to return an error if you try to call the function more than once ( `TODO: return an error and finish the stream instead, due to API misuse?` )... and maybe possibly in other cases? I'm not sure on the intentions behind always just returning `nil` instead of not returning anything.

I'll add a comment here.